### PR TITLE
Improved Unit Test coverage and scenarios

### LINKED ISSUE #206: Refactor test API directory to support versioned structure (`/tests/memmachine/server/`)
Refactor the test directory structure for API tests to improve maintainability and clarity as new API versions are introduced.

## Details
- Move existing tests from `tests/memmachine/server/` into versioned subdirectories:
  - For current tests, place under `tests/memmachine/server/v1/`
  - For future API versions, create corresponding directories (e.g., `v2`, `v3`) as needed
- Update import paths and fixtures to support the new structure
- Ensure all test discovery and CI config continue to work after the refactor
- Document the new organization in the test directory's README or project documentation

Proposed directory structure

```text
tests/
  memmachine/
    server/
      v1/
        test-rest-memories.py
        test-other-endpoint.py
      v2/
        test-rest-memories.py
        test-other-endpoint.py
      # ... future versions
```

## Motivation
- Isolate tests for different API versions to prevent confusion and mixing of version-specific behaviors
- Make it easier to deprecate, update, or compare tests across versions
- Improve onboarding for new contributors by providing a clear organizational structure

## Acceptance Criteria
- All existing tests for `/v1` endpoints are moved to the new versioned directory
- Project documentation is updated to reflect the new structure
- CI passes after the directory changes

## Additional context
This refactoring will help keep the codebase maintainable as the API evolves and new versions are released.

### LINKED ISSUE #208: [Bug]: Unit tests for `tests/memmachine/server/test_rest_memories.py` are incomplete
### Describe the bug

As shown in #202, the unit tests don't cover all the possible scenarios.

### Steps to reproduce

Running `pytest tests/memmachine/server` shows all tests completed without error.

### Expected behavior

The unit tests should have found the issue described in #202 

### Environment

- OS: Ubuntu Linux 25.04
- MemMachine v0.1.4

### Additional context

The unit tests don't provide full coverage for the DELETE operations, and we're missing some scenarios.

### PR DESCRIPTION
**Pull request by sscargal**
### Purpose of the change

The coverage of the tests in `test_rest_memories.py` was not extensive enough. Added more coverage and scenarios.

### Description

#202 was filed showing a gap in the unit testing. This PR now fixes this scenario and adds a lot more.

================================================================================
MemMachine REST API /v1/memories Endpoint Tests

This test suite provides comprehensive coverage for the /v1/memories endpoint,
including POST, DELETE, and query operations. It validates various scenarios:

- Valid payloads (string, list content for embeddings)
- Missing required fields (top-level and nested session fields)
- Invalid data types for fields
- Handling of extra, unexpected fields
- Boundary conditions for empty values and null metadata
- Rejection of null values for required session fields
- Robustness against malicious/junk input (path traversal, null byte injection, SQL injection)
- Numerical boundary conditions for 'episode_content' (embeddings), covering
  negative, zero, large, fractional numbers, and rejection of special floats (inf, nan).
- Rejection of unsupported HTTP methods (e.g., GET).

The tests utilize mocking of AsyncEpisodicMemory to isolate endpoint logic.
================================================================================

### Fixes/Closes

Fixes #208
Fixes #206

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration.

- [x] Unit Test
- [x] Manual verification (list step-by-step instructions)

**Test Results:** 

The tests show several new issues to be resolved

```
$ pytest v1/
=================================================================================== test session starts ===================================================================================
platform linux -- Python 3.13.3, pytest-8.4.2, pluggy-1.6.0
rootdir: /home/steve/MemMachine
configfile: pyproject.toml
plugins: asyncio-1.2.0, anyio-4.10.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 69 items                                                                                                                                                                        

v1/test_rest_health_endpoint.py ..                                                                                                                                                  [  2%]
v1/test_rest_memories.py .................FFF..FFFFFFFF.......FFF.............FFF..FFFFFFFF.                 

--- 8< ---

================================================================================= short test summary info =================================================================================
FAILED v1/test_rest_memories.py::test_post_memories_none_session_field[group_id] - AssertionError: Should fail with None session field: group_id
FAILED v1/test_rest_memories.py::test_post_memories_none_session_field[agent_id] - AssertionError: Should fail with None session field: agent_id
FAILED v1/test_rest_memories.py::test_post_memories_none_session_field[user_id] - AssertionError: Should fail with None session field: user_id
FAILED v1/test_rest_memories.py::test_post_memories_malicious_input[group_id-../../../../../../../../../../../../../../../../../etc/passwd] - AssertionError: Should fail with malicious input in group_id: ../../../../../../../../../../../../../../../../../etc/passwd
FAILED v1/test_rest_memories.py::test_post_memories_malicious_input[group_id-valid_id\\u0000malicious_suffix] - AssertionError: Should fail with malicious input in group_id: valid_id\u0000malicious_suffix
FAILED v1/test_rest_memories.py::test_post_memories_malicious_input[group_id-' OR '1'='1 --] - AssertionError: Should fail with malicious input in group_id: ' OR '1'='1 --
FAILED v1/test_rest_memories.py::test_post_memories_malicious_input[session_id-../../../../../../../../../../../../../../../../../etc/passwd] - AssertionError: Should fail with malicious input in session_id: ../../../../../../../../../../../../../../../../../etc/passwd
FAILED v1/test_rest_memories.py::test_post_memories_malicious_input[session_id-valid_session\\u0000malicious_suffix] - AssertionError: Should fail with malicious input in session_id: valid_session\u0000malicious_suffix
FAILED v1/test_rest_memories.py::test_post_memories_malicious_input[session_id-' OR '1'='1 --] - AssertionError: Should fail with malicious input in session_id: ' OR '1'='1 --
FAILED v1/test_rest_memories.py::test_post_memories_malicious_input[agent_id-malicious_value6] - AssertionError: Should fail with malicious input in agent_id: ['agent1', '../../../etc/shadow']
FAILED v1/test_rest_memories.py::test_post_memories_malicious_input[user_id-malicious_value7] - AssertionError: Should fail with malicious input in user_id: ['user1', 'user2\\u0000evil']
FAILED v1/test_rest_memories.py::test_post_memories_numerical_boundaries_episode_content[numerical_list7-422] - AssertionError: Expected status 422 for numerical list [inf], got 200
FAILED v1/test_rest_memories.py::test_post_memories_numerical_boundaries_episode_content[numerical_list8-422] - AssertionError: Expected status 422 for numerical list [-inf], got 200
FAILED v1/test_rest_memories.py::test_post_memories_numerical_boundaries_episode_content[numerical_list9-422] - AssertionError: Expected status 422 for numerical list [nan], got 200
FAILED v1/test_rest_memories.py::test_delete_memories_none_session_field[group_id] - AssertionError: Should fail with None session field: group_id
FAILED v1/test_rest_memories.py::test_delete_memories_none_session_field[agent_id] - AssertionError: Should fail with None session field: agent_id
FAILED v1/test_rest_memories.py::test_delete_memories_none_session_field[user_id] - AssertionError: Should fail with None session field: user_id
FAILED v1/test_rest_memories.py::test_delete_memories_malicious_input[group_id-../../../../../../../../../../../../../../../../../etc/passwd] - AssertionError: Should fail with malicious input in group_id: ../../../../../../../../../../../../../../../../../etc/passwd
FAILED v1/test_rest_memories.py::test_delete_memories_malicious_input[group_id-valid_id\\u0000malicious_suffix] - AssertionError: Should fail with malicious input in group_id: valid_id\u0000malicious_suffix
FAILED v1/test_rest_memories.py::test_delete_memories_malicious_input[group_id-' OR '1'='1 --] - AssertionError: Should fail with malicious input in group_id: ' OR '1'='1 --
FAILED v1/test_rest_memories.py::test_delete_memories_malicious_input[session_id-../../../../../../../../../../../../../../../../../etc/passwd] - AssertionError: Should fail with malicious input in session_id: ../../../../../../../../../../../../../../../../../etc/passwd
FAILED v1/test_rest_memories.py::test_delete_memories_malicious_input[session_id-valid_session\\u0000malicious_suffix] - AssertionError: Should fail with malicious input in session_id: valid_session\u0000malicious_suffix
FAILED v1/test_rest_memories.py::test_delete_memories_malicious_input[session_id-' OR '1'='1 --] - AssertionError: Should fail with malicious input in session_id: ' OR '1'='1 --
FAILED v1/test_rest_memories.py::test_delete_memories_malicious_input[agent_id-malicious_value6] - AssertionError: Should fail with malicious input in agent_id: ['agent1', '../../../etc/shadow']
FAILED v1/test_rest_memories.py::test_delete_memories_malicious_input[user_id-malicious_value7] - AssertionError: Should fail with malicious input in user_id: ['user1', 'user2\\u0000evil']
============================================================================== 25 failed, 44 passed in 1.38s ==============================================================================
```

### Checklist

- [x] I have signed the commit(s) within this pull request
- [x] My code follows the style guidelines of this project (See STYLE_GUIDE.md)
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [x] My changes generate no new warnings
- [x] I have added unit tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have checked my code and corrected any misspellings

### Maintainer Checklist

- [ ] Confirmed all checks passed
- [ ] Contributor has signed the commit(s)
- [ ] Reviewed the code
- [ ] Run, Tested, and Verified the change(s) work as expected

### Screenshots/Gifs

See output above

### Further comments

New bugs need to be filed based on the output of these new tests.

### CODE REVIEW

File: tests/memmachine/server/v1/test_rest_memories.py [hunk 1]
```diff
@@ -0,0 +1,394 @@
...
+    "/v1/memories",
+    "/v1/memories/episodic",
+    "/v1/memories/profile"
+])
+def test_post_valid_content_to_endpoints(valid_post_payload, endpoint):
+    """Tests successful memory ingestion across all relevant POST endpoints."""
+    payload = valid_post_payload.copy() # Use a copy to avoid modifying fixture for other tests
+    if "profile" in endpoint:
+        payload["episode_type"] = "embedding"
+        payload["episode_content"] = [0.1, 0.2] # Profile usually expects embeddings
```
Comment (edwinyyyu) [inline tests/memmachine/server/v1/test_rest_memories.py]: @malatewang Is this how the profile memory is supposed to behave?

File: tests/memmachine/server/v1/test_rest_memories.py [hunk 2]
```diff
@@ -0,0 +1,394 @@
...
+    response = client.post(endpoint, json=payload)
+    assert response.status_code in (200, 201, 204)
+
+
+def test_post_memories_valid_list_content(valid_post_payload):
+    """Tests successful memory ingestion with a valid list (embedding)."""
+    valid_post_payload["episode_content"] = [0.1, 0.2, 0.3]
+    valid_post_payload["episode_type"] = "embedding"
+    response = client.post("/v1/memories", json=valid_post_payload)
+    assert response.status_code in (200, 201, 204)
```
Comment (edwinyyyu) [inline tests/memmachine/server/v1/test_rest_memories.py]: How is this test useful vs. the previous test?

Also, while technically it's okay, we don't actually support ingesting embeddings anywhere as far as I know.

File: tests/memmachine/server/v1/test_rest_memories.py [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+    ("group_id", "../../../../../../../../../../../../../../../../../etc/passwd"),
+    ("group_id", "valid_id\\u0000malicious_suffix"),
+    ("group_id", "' OR '1'='1 --"),
+    ("session_id", "../../../../../../../../../../../../../../../../../etc/passwd"),
+    ("session_id", "valid_session\\u0000malicious_suffix"),
+    ("session_id", "' OR '1'='1 --"),
+    ("agent_id", ["agent1", "../../../etc/shadow"]),
+    ("user_id", ["user1", "user2\\u0000evil"])
+])
+def test_post_memories_malicious_input(valid_post_payload, field, malicious_value):
```
Comment (edwinyyyu) [inline tests/memmachine/server/v1/test_rest_memories.py:284]: Do we have actual logic to check these?

File: tests/memmachine/server/v1/test_rest_memories.py [hunk 4]
```diff
@@ -0,0 +1,394 @@
...
+    ([float('inf')], 422),
+    ([float('-inf')], 422),
+    ([float('nan')], 422),
+    (["not_a_number"], 422), # Non-numerical item in list
+    ([1, "invalid", 3], 422), # Mixed valid and invalid
+])
+def test_post_memories_numerical_boundaries_episode_content(valid_post_payload, numerical_list, expected_status):
+    """
+    Tests POST requests with various numerical boundary conditions for episode_content (embeddings).
+    Checks for valid float lists and rejection of invalid float values like inf, -inf, nan.
```
Comment (edwinyyyu) [inline tests/memmachine/server/v1/test_rest_memories.py:326]: Why not support these? These are standard IEEE float values I think. This makes it harder for the user to input their data.

File: tests/memmachine/server/v1/test_rest_memories.py [hunk 5]
```diff
@@ -0,0 +1,394 @@
...
+    ("group_id", "../../../../../../../../../../../../../../../../../etc/passwd"),
+    ("group_id", "valid_id\\u0000malicious_suffix"),
+    ("group_id", "' OR '1'='1 --"),
+    ("session_id", "../../../../../../../../../../../../../../../../../etc/passwd"),
+    ("session_id", "valid_session\\u0000malicious_suffix"),
+    ("session_id", "' OR '1'='1 --"),
+    ("agent_id", ["agent1", "../../../etc/shadow"]),
+    ("user_id", ["user1", "user2\\u0000evil"])
+])
+def test_delete_memories_malicious_input(valid_delete_payload, field, malicious_value):
```
Comment (edwinyyyu) [inline tests/memmachine/server/v1/test_rest_memories.py:494]: Do we have actual logic to check these? Like checking for alphanumeric only?

File: tests/memmachine/server/v1/test_rest_memories.py (hunk 1 above)
Comment (sscargal) [reply tests/memmachine/server/v1/test_rest_memories.py]: I checked the code, and you are correct. We do not support passing embeddings directly through the API. I will remove this test. It tests multiple endpoints under `/memories`. My plan is to have a dedicated test file for each endpoint, so this test will be duplicative across test files.

File: tests/memmachine/server/v1/test_rest_memories.py (hunk 2 above)
Comment (sscargal) [reply tests/memmachine/server/v1/test_rest_memories.py]: I will remove this test as we do not support passing embeddings.

File: tests/memmachine/server/v1/test_rest_memories.py (hunk 5 above)
Comment (sscargal) [reply tests/memmachine/server/v1/test_rest_memories.py:494]: There is currently no explicit logic to check for "alphanumeric only" or to specifically reject malicious characters like path traversal sequences (`../../../etc/passwd`) or null bytes (`\u0000`) within these fields in `app.py`.

Pydantic's default validation for str types ensures that the input is indeed a string. However, it does not, by default, perform content-based validation (e.g., checking for specific character sets, preventing null bytes, or disallowing path traversal patterns).

The application (app.py) accepted the string containing the null byte as a valid str, indicating a lack of specific content validation for such inputs.

To implement such checks, we would need to add explicit validation rules to the SessionData Pydantic model, for example, using pydantic.Field with regex patterns or custom validators.

File: tests/memmachine/server/v1/test_rest_memories.py (hunk 3 above)
Comment (sscargal) [reply tests/memmachine/server/v1/test_rest_memories.py:284]: This has been implemented in `app.py` now as part of this PR.

File: tests/memmachine/server/v1/test_rest_memories.py (hunk 4 above)
Comment (sscargal) [reply tests/memmachine/server/v1/test_rest_memories.py:326]: This test now sends list[float] values (which are numerical lists) for episode_content and correctly asserts that the API returns a 422 Unprocessable Entity status code. This confirms that the API is rejecting non-string inputs for episode_content, which aligns with the updated Pydantic model in `app.py`.

Regarding "harder for the user to input their data":

Yes, if users need to input numerical data (like what would typically be an embedding) into the episode_content field, they will now have to serialize that numerical data into a string before sending it to the API. For example, a list of floats [0.1, 0.2, 0.3] would need to be sent as the string "[0.1, 0.2, 0.3]".

The API will then treat and store this as a regular string. Any downstream processing that requires the numerical nature of this data would need to parse the string back into a numerical format. This is a direct consequence of the design choice not to support direct embeddings (i.e., list[float]) for episode_content.

So, the test is correctly verifying the API's adherence to the new design, which inherently makes inputting structured numerical data more manual for the user.

**Comment by sscargal** [comment]
Thanks for the feedback. With the changes to the unit test and app.py, the number of failing tests has been reduced to 16

```
================================================================================= short test summary info =================================================================================
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_post_memories_missing_nested_session_field[group_id] - AssertionError: Should fail with missing session field: group_id
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_post_memories_missing_nested_session_field[agent_id] - AssertionError: Should fail with missing session field: agent_id
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_post_memories_missing_nested_session_field[user_id] - AssertionError: Should fail with missing session field: user_id
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_post_memories_boundary_empty_values - assert 422 in (200, 201, 204)
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_post_memories_none_session_field[group_id] - AssertionError: Should fail with None session field: group_id
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_post_memories_none_session_field[agent_id] - AssertionError: Should fail with None session field: agent_id
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_post_memories_none_session_field[user_id] - AssertionError: Should fail with None session field: user_id
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_post_memories_numerical_boundaries_episode_content[numerical_list7-422] - AssertionError: Expected status 422 for numerical list [inf], got 500
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_post_memories_numerical_boundaries_episode_content[numerical_list8-422] - AssertionError: Expected status 422 for numerical list [-inf], got 500
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_post_memories_numerical_boundaries_episode_content[numerical_list9-422] - AssertionError: Expected status 422 for numerical list [nan], got 500
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_delete_memories_missing_nested_session_field[group_id] - AssertionError: Should fail with missing session field: group_id
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_delete_memories_missing_nested_session_field[agent_id] - AssertionError: Should fail with missing session field: agent_id
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_delete_memories_missing_nested_session_field[user_id] - AssertionError: Should fail with missing session field: user_id
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_delete_memories_none_session_field[group_id] - AssertionError: Should fail with None session field: group_id
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_delete_memories_none_session_field[agent_id] - AssertionError: Should fail with None session field: agent_id
FAILED tests/memmachine/server/v1/test_rest_memories.py::test_delete_memories_none_session_field[user_id] - AssertionError: Should fail with None session field: user_id
======================================================================== 16 failed, 48 passed, 1 warning in 1.30s =========================================================================
```

File: tests/memmachine/server/v1/test_rest_memories.py (hunk 4 above)
Comment (sscargal) [reply tests/memmachine/server/v1/test_rest_memories.py:326]: I implemented the fixes.

File: tests/memmachine/server/v1/test_rest_memories.py (hunk 5 above)
Comment (sscargal) [reply tests/memmachine/server/v1/test_rest_memories.py:494]: I implemented the fixes.

**Comment by sscargal** [comment]
I added the fixes to `app.py` and updated the unit tests, so they all complete successfully now

```
$ pytest tests/memmachine/server/
=================================================================================== test session starts ===================================================================================
platform linux -- Python 3.13.3, pytest-8.4.2, pluggy-1.6.0
rootdir: /home/steve/MemMachine
configfile: pyproject.toml
plugins: asyncio-1.2.0, anyio-4.10.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 66 items                                                                                                                                                                        

tests/memmachine/server/v1/test_rest_health_endpoint.py ..                                                                                                                          [  3%]
tests/memmachine/server/v1/test_rest_memories.py ................................................................                                                                   [100%]

=================================================================================== 66 passed in 1.21s ====================================================================================
```

File: src/memmachine/server/app.py [hunk 6]
```diff
@@ -176,7 +200,7 @@ async def http_app_lifespan(application: FastAPI):
 
     # TODO switch to using builder initialization
     llm_model = OpenAILanguageModel({"api_key": api_key, "model": model})
-    embeddings = OpenAIEmbedder({"api_key": api_key})
+    embeddings = OpenAIEmbedder({"api_key": api_key})  # noqa: F821
```
Comment (edwinyyyu) [inline src/memmachine/server/app.py]: What is the `#noqa` doing here?
If it needs the embedder, import the embedder.
If it does not need the embedder, do not create the embedder.

**Comment by sscargal** [comment]
This needs significant reworking after a few changes were merged.

File: src/memmachine/server/app.py (hunk 6 above)
Comment (sscargal) [reply src/memmachine/server/app.py]: Removed.

**Comment by sscargal** [comment]
I fixed the newer reported bugs in `app.py` and added new unit test coverage for them. All tests pass cleanly.

```
$ pytest tests/memmachine/server/
=================================================================================================== test session starts ===================================================================================================
platform linux -- Python 3.13.3, pytest-8.4.2, pluggy-1.6.0
rootdir: /home/steve/MemMachine
configfile: pyproject.toml
plugins: asyncio-1.2.0, anyio-4.10.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 91 items                                                                                                                                                                                                        

tests/memmachine/server/v1/test_rest_health_endpoint.py ..                                                                                                                                                          [  2%]
tests/memmachine/server/v1/test_rest_memories.py .........................................................................................                                                                          [100%]

=================================================================================================== 91 passed in 1.27s ====================================================================================================
```

File: src/memmachine/server/app.py [hunk 7]
```diff
@@ -48,31 +50,66 @@
...
+
+
+EpisodeContentString = Annotated[str, BeforeValidator(validate_episode_content_type)]
+
+
+class EpisodeType(str, Enum):
+    """Enum for the types of episodes that can be stored."""
+
+    MESSAGE = "message"
+    EMBEDDING = "embedding"
```
Comment (edwinyyyu) [inline src/memmachine/server/app.py]: I think we don't support anything except for string content type, which would be for episode types such as message right now?

Is there a reason to include embedding as an episode type?

**Comment by sscargal** [comment]
I implemented the previous feedback

```
$ pytest tests/memmachine/server/
=================================================================================================== test session starts ====================================================================================================
platform linux -- Python 3.13.3, pytest-8.4.2, pluggy-1.6.0
rootdir: /home/steve/MemMachine
configfile: pyproject.toml
plugins: asyncio-1.2.0, anyio-4.10.0
asyncio: mode=Mode.STRICT, debug=False, asyncio_default_fixture_loop_scope=None, asyncio_default_test_loop_scope=function
collected 80 items                                                                                                                                                                                                         

tests/memmachine/server/v1/test_rest_health_endpoint.py ..                                                                                                                                                           [  2%]
tests/memmachine/server/v1/test_rest_memories.py ..............................................................................                                                                                      [100%]

==================================================================================================== 80 passed in 1.25s ====================================================================================================
```

**Comment by sscargal** [comment]
Closing as there are now unit tests merged for the REST API endpoints.
